cli/azd/pkg/input/previewer.go

```diff
@@ -0,0 +1,296 @@
+type progressLog struct {
+	// The initial top wording. This value can be updated using Header() method.
+	header string
+	// The raw initial title for the component. This value can't be changed after Starting the component.
+	title string
+	// The title that is displayed on the form of `<prefix>──── <title> ──────────────`.
+	displayTitle string
+	// The line on the bottom. The value is saved on the component state so it is not generated on every re-draw.
+	footerLine string
+	// The number of rows to display for logging.
+	lines int
+	// Use prefix for indentation or any other symbol before each line.
+	prefix string
+	// This list is used as the memory buffer for the logs. The buffer is kept with the size of `lines`
+	output []string
+	// The mutex is used to coordinate updating the header, stopping the component and printing logs.
+	outputMutex sync.Mutex
+}
+
+/****************** Exported method ****************
+- NewProgressLog
+- Start
+- Stop
+- Write
+- Header
+*/
+
+// NewProgressLog returns a new instance of a progressLog.
+func NewProgressLog(lines int, prefix, title, header string) *progressLog {
+	return &progressLog{
+		lines:  lines,
+		prefix: prefix,
+		title:  title,
+		header: header,
+	}
+}
+
+// Start must be call to initialize the component and draw the initial empty frame in the screen.
+// Calling Start() a second time is a no-op. The screen is only updated during the first call.
+// Call Stop() to remove the frame and logs from the screen.
+func (p *progressLog) Start() {
+	if p.output != nil {
+		return
+	}
+	p.output = make([]string, p.lines)
+	// title is created on Start() because it depends on terminal width
+	// if terminal is resized between stop and start, the previewer will
+	// react to it and update the size.
+	p.buildTopBottom()
+
+	// header
+	if p.header != "" {
+		tm.ResetLine("")
+		tm.Println(p.header)
+		// margin after title
+		tm.Println("")
+	}
+
+	// title
+	tm.ResetLine("")
+	tm.Println(p.displayTitle)
+	tm.Println("")
+
+	p.printLogs()
+}
+
+// Stop clears the screen from any previous output and clear the buffer.
+// Calling Stop() before Start() is a no-op.
+func (p *progressLog) Stop() {
+	if p.output == nil {
+		return
+	}
+
+	p.outputMutex.Lock()
+	defer p.outputMutex.Unlock()
+
+	p.clearContentAndFlush()
+
+	// title
+	tm.MoveCursorUp(2)
+	clearLine(p.displayTitle)
+
+	// header
+	if p.header != "" {
+		tm.MoveCursorUp(2)
+		clearLine(p.header)
+	}
+
+	tm.Flush()
+	p.output = nil
+}
+
+// Write implements oi.Writer and updates the internal buffer before flushing it into the screen.
+// Calling Write() before Start() or after Stop() is a no-op
+func (p *progressLog) Write(logBytes []byte) (int, error) {
+	if p.output == nil {
+		return len(logBytes), nil
+	}
+	logsScanner := bufio.NewScanner(strings.NewReader(string(logBytes)))
+	maxWidth := tm.Width()
+	if maxWidth <= 0 {
+		// tm.Width <= 0 means there's no terminal to write and the stdout pipe is mostly connected to a file or a buffer
```

Comment:
Have a feeling that we might want to do this at a higher level.

Suppose that we now use `console.ShowPreviewer` which now hijacks the `console.Writer`, and `tm.Width()<0`, this ends up just discarding output?

Reply:
discards only the output for the logProgress component, which has no meaning while running without a terminal